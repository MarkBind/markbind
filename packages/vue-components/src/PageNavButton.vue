<template>
  <overlay
    v-if="showPageNav"
    type="pageNav"
    :portal-name="portalName"
  >
    <template #navMenuIcon>
      <div :class="['toggle-page-nav-button']">
        <span></span>
        <span></span>
        <span></span>
      </div>
    </template>
  </overlay>
</template>

<script>
import $ from './utils/NodeList';

export default {
  data() {
    return {
      portalName: undefined,
      show: false,
    };
  },
  computed: {
    showPageNav() {
      return this.show && this.portalName;
    },
  },
  methods: {
    togglePageNavButton() {
      if (window.innerWidth < 1300) {
        this.show = true;
      } else {
        this.show = false;
      }
    },
  },
  mounted() {
    if (document.querySelector('#page-nav a') !== null) {
      this.portalName = 'page-nav';
    } else if (document.querySelector('#mb-page-nav a') !== null) {
      this.portalName = 'mb-page-nav';
    }

    this.togglePageNavButton();
    $(window).on('resize', this.togglePageNavButton);
  },
  beforeUnmount() {
    $(window).off('resize', this.togglePageNavButton);
  },
};
</script>

<style scoped>
    .toggle-page-nav-button {
        cursor: pointer;
        top: 6px;
        padding: 15px;
        position: absolute;
        right: 0;
    }

    .toggle-page-nav-button span {
        background-color: #000;
        border-radius: 50%;
        display: block;
        height: 5px;
        margin-top: 2px;
        position: absolute;
        right: 15px;
        transform: rotate(0deg);
        transition: all 0.25s ease-in-out;
        width: 5px;
    }

    .toggle-page-nav-button > span:nth-child(1) {
        transform: translateY(-8px);
    }

    .toggle-page-nav-button > span:nth-child(2) {
        transform-origin: 100% 50%;
    }

    .toggle-page-nav-button > span:nth-child(3) {
        transform: translateY(8px);
    }

    .nav-menu-close-icon .toggle-page-nav-button > span:nth-child(1) {
        border-radius: 0;
        height: 3px;
        transform: rotate(135deg);
        width: 21px;
    }

    .nav-menu-close-icon .toggle-page-nav-button > span:nth-child(2) {
        transform: scale(0);
    }

    .nav-menu-close-icon .toggle-page-nav-button > span:nth-child(3) {
        border-radius: 0;
        height: 3px;
        transform: rotate(-135deg);
        width: 21px;
    }

</style>
