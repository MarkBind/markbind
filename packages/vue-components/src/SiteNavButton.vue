<template>
  <overlay
    v-if="showSiteNav"
    type="siteNav"
    :portal-name="portalName"
  >
    <template #navMenuIcon>
      <div
        :class="['toggle-site-nav-button']"
      >
        <span></span>
        <span></span>
        <span></span>
      </div>
    </template>
  </overlay>
</template>

<script>
import $ from './utils/NodeList';

export default {
  data() {
    return {
      portalName: undefined,
      show: false,
    };
  },
  computed: {
    showSiteNav() {
      return this.show && this.portalName;
    },
  },
  methods: {
    toggleSiteNavButton() {
      if (window.innerWidth < 992) {
        this.show = true;
      } else {
        this.show = false;
      }
    },
  },
  mounted() {
    if (document.querySelector('#site-nav a') !== null) {
      this.portalName = 'site-nav';
    } else if (document.querySelector('.site-nav-root a') !== null) {
      this.portalName = 'mb-site-nav';
    }

    this.toggleSiteNavButton();
    $(window).on('resize', this.toggleSiteNavButton);
  },
  beforeUnmount() {
    $(window).off('resize', this.toggleSiteNavButton);
  },
};
</script>

<style scoped>

    .toggle-site-nav-button {
        cursor: pointer;
        height: 53px;
        left: 0;
        padding: 15px;
        position: absolute;
        transform: rotate(0deg);
        transition: all 0.25s ease-in-out;
        width: 52px;
    }

    .toggle-site-nav-button span {
        background-color: #000;
        display: block;
        height: 3px;
        left: 15px;
        position: absolute;
        top: 50%;
        transition: all 0.25s ease-in-out;
        width: 21px;
    }

    .toggle-site-nav-button > span:nth-child(1) {
        transform: translateY(-8px);
    }

    .toggle-site-nav-button > span:nth-child(2) {
        transform-origin: 100% 50%;
    }

    .toggle-site-nav-button > span:nth-child(3) {
        transform: translateY(8px);
    }

    .nav-menu-close-icon .toggle-site-nav-button > span:nth-child(1) {
        transform: rotate(135deg);
    }

    .nav-menu-close-icon .toggle-site-nav-button > span:nth-child(2) {
        left: -60px;
        opacity: 0;
    }

    .nav-menu-close-icon .toggle-site-nav-button > span:nth-child(3) {
        transform: rotate(-135deg);
    }

</style>
