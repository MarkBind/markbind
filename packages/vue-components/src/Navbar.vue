<template>
  <div>
    <nav ref="navbar" :class="['navbar', 'navbar-expand-md', themeOptions, addClass, fixedOptions]">
      <div class="container-fluid">
        <div class="navbar-brand">
          <slot name="brand"></slot>
        </div>
        <button
          v-if="!slots.collapse"
          class="navbar-toggler"
          type="button"
          aria-expanded="false"
          aria-label="Toggle navigation"
          @click="toggleCollapse"
        >
          <span class="navbar-toggler-icon"></span>
          <slot name="collapse"></slot>
        </button>

        <div :class="['navbar-collapse',{collapse:collapsed}]">
          <ul class="navbar-nav mr-auto mt-2 mt-lg-0">
            <slot></slot>
          </ul>
          <ul v-if="slots.right" class="navbar-nav navbar-right">
            <slot name="right"></slot>
          </ul>
        </div>
      </div>
    </nav>
    <div
      v-show="isLowerNavbarShowing"
      ref="lowerNavbar"
      class="lower-navbar-container"
    >
      <slot name="lower-navbar">
        <site-nav-button />
        <page-nav-button />
      </slot>
    </div>
  </div>
</template>

<script>
import $ from './utils/NodeList';
import { toBoolean } from './utils/utils';
import normalizeUrl from './utils/urls';
import SiteNavButton from './SiteNavButton.vue';
import PageNavButton from './PageNavButton.vue';

export default {
  components: {
    SiteNavButton,
    PageNavButton,
  },
  props: {
    type: {
      type: String,
      default: 'primary',
    },
    addClass: {
      type: String,
      default: '',
    },
    fixed: {
      type: Boolean,
      default: false,
    },
    defaultHighlightOn: {
      type: String,
      default: 'sibling-or-child',
    },
  },
  provide() {
    return {
      toggleLowerNavbar: this.toggleLowerNavbar,
    };
  },
  data() {
    return {
      id: 'bs-example-navbar-collapse-1',
      collapsed: true,
      styles: {},
      isLowerNavbarShowing: false,
    };
  },
  computed: {
    fixedBool() {
      return toBoolean(this.fixed);
    },
    fixedOptions() {
      if (this.fixedBool) {
        return 'navbar-fixed';
      }
      return '';
    },
    slots() {
      return this.$scopedSlots;
    },
    themeOptions() {
      switch (this.type) {
      case 'none':
        return '';
      case 'light':
        return 'navbar-light bg-light';
      case 'dark':
        return 'navbar-dark bg-dark';
      case 'primary':
      default:
        return 'navbar-dark bg-primary';
      }
    },
  },
  methods: {
    toggleCollapse(e) {
      if (e) { e.preventDefault(); }
      this.collapsed = !this.collapsed;
    },
    // Splits a normalised URL into its parts,
    // e.g http://site.org/foo/bar/index.html -> ['foo','bar','index.html']
    splitUrl(url) {
      const u = new URL(normalizeUrl(url));
      return `${u.pathname}`.substr(1).split('/');
    },
    isSibling(url, href) {
      const hParts = this.splitUrl(href);
      const uParts = this.splitUrl(url);
      if (hParts.length !== uParts.length) {
        return false;
      }
      for (let i = 0; i < hParts.length - 1; i += 1) {
        if (hParts[i] !== uParts[i]) {
          return false;
        }
      }
      return true;
    },
    isChild(url, href) {
      const hParts = this.splitUrl(href);
      const uParts = this.splitUrl(url);
      if (uParts.length <= hParts.length) {
        return false;
      }
      for (let i = 0; i < hParts.length; i += 1) {
        if (hParts[i] !== uParts[i]) {
          return false;
        }
      }
      return true;
    },
    isExact(url, href) {
      return normalizeUrl(url) === normalizeUrl(href);
    },
    addClassIfDropdown(dropdownLinks, a, li) {
      if (dropdownLinks.includes(a)) {
        a.classList.add('dropdown-current');
        this.addClassIfSubmenu(a, li);
      }
    },
    addClassIfSubmenu(a, li) {
      let el = a.parentElement;
      while (el !== li) {
        if (el.classList.contains('dropdown-submenu')) {
          $(el).findChildren('a').each(aChild => aChild.classList.add('dropdown-current'));
        }
        el = el.parentElement;
      }
    },
    highlightLink(url) {
      const defHlMode = this.defaultHighlightOn;
      const navLis = Array.from(this.$el.querySelector('.navbar-nav').children);
      // attempt an exact match first
      for (let i = 0; i < navLis.length; i += 1) {
        const li = navLis[i];
        const standardLinks = [li];
        const navLinks = Array.from(li.querySelectorAll('a.nav-link'));
        const dropdownLinks = Array.from(li.querySelectorAll('a.dropdown-item'));
        const allNavLinks = standardLinks.concat(navLinks).concat(dropdownLinks).filter(a => a.href);
        for (let j = 0; j < allNavLinks.length; j += 1) {
          const a = allNavLinks[j];
          const hlMode = a.getAttribute('highlight-on') || defHlMode;
          if (hlMode === 'none') {
            // eslint-disable-next-line no-continue
            continue;
          }
          // terminate early on an exact match
          if (this.isExact(url, a.href)) {
            li.classList.add('current');
            this.addClassIfDropdown(dropdownLinks, a, li);
            return;
          }
        }
      }
      // fallback to user preference, otherwise
      for (let i = 0; i < navLis.length; i += 1) {
        const li = navLis[i];
        const standardLinks = [li];
        const navLinks = Array.from(li.querySelectorAll('a.nav-link'));
        const dropdownLinks = Array.from(li.querySelectorAll('a.dropdown-item'));
        const allNavLinks = standardLinks.concat(navLinks).concat(dropdownLinks).filter(a => a.href);
        for (let j = 0; j < allNavLinks.length; j += 1) {
          const a = allNavLinks[j];
          const hlMode = a.getAttribute('highlight-on') || defHlMode;
          if (hlMode === 'none') {
            // eslint-disable-next-line no-continue
            continue;
          }
          // Ignores invalid navbar highlight rule
          if (hlMode === 'sibling-or-child') {
            if (this.isSibling(url, a.href) || this.isChild(url, a.href)) {
              li.classList.add('current');
              this.addClassIfDropdown(dropdownLinks, a, li);
              return;
            }
          } else if (hlMode === 'sibling') {
            if (this.isSibling(url, a.href)) {
              li.classList.add('current');
              this.addClassIfDropdown(dropdownLinks, a, li);
              return;
            }
          } else if (hlMode === 'child') {
            if (this.isChild(url, a.href)) {
              li.classList.add('current');
              this.addClassIfDropdown(dropdownLinks, a, li);
              return;
            }
          }
        }
      }
    },
    toggleLowerNavbar() {
      if (this.$refs.lowerNavbar.childElementCount > 0) {
        this.isLowerNavbarShowing = true;
      } else {
        this.isLowerNavbarShowing = false;
      }
    },
  },
  created() {
    this._navbar = true;
  },
  mounted() {
    const $dropdown = $('.dropdown>[data-toggle="dropdown"]', this.$el).parent();
    $dropdown.on('click', '.dropdown-toggle', (e) => {
      e.preventDefault();
      $dropdown.each((content) => {
        if (content.contains(e.target)) content.classList.toggle('open');
      });
    }).on('click', '.dropdown-menu>li>a', (e) => {
      $dropdown.each((content) => {
        if (content.contains(e.target)) content.classList.remove('open');
      });
    }).onBlur((e) => {
      $dropdown.each((content) => {
        if (!content.contains(e.target)) content.classList.remove('open');
      });
    });
    $(this.$el).on('click', 'li:not(.dropdown)>a', (e) => {
      if (e.target.classList.contains('submenu-toggle')) { return; }
      setTimeout(() => { this.collapsed = true; }, 200);
    }).onBlur((e) => {
      if (!this.$el.contains(e.target)) { this.collapsed = true; }
    });
    if (this.slots.collapse) $('[data-toggle="collapse"]', this.$el).on('click', e => this.toggleCollapse(e));

    // highlight current nav link
    this.highlightLink(window.location.href);

    this.toggleLowerNavbar();
    $(window).on('resize', this.toggleLowerNavbar);
  },
  beforeDestroy() {
    $('.dropdown', this.$el).off('click').offBlur();
    if (this.slots.collapse) $('[data-toggle="collapse"]', this.$el).off('click');
    $(window).off('resize', this.toggleLowerNavbar);
  },
};
</script>

<style scoped>
    @media (max-width: 767px) {
        .navbar-collapse {
            max-height: 80vh !important;
            overflow-x: hidden !important;
            overflow-y: scroll !important;
        }
    }

    .navbar-fixed {
        position: fixed;
        width: 100%;
        z-index: 1000;
    }

    >>> .dropdown-current {
        color: #fff !important;
        background: #007bff;
    }

    .lower-navbar-container {
        background-color: #fff;
        border-bottom: 1px solid #c1c1c1;
        height: 50px;
        width: 100%;
        position: relative;
    }
</style>
