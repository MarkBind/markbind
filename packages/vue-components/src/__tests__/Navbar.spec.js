import { mount } from '@vue/test-utils';
import Navbar from '../Navbar.vue';
import PageNavButton from '../PageNavButton.vue';
import SiteNavButton from '../SiteNavButton.vue';

const DEFAULT_STUBS = {
  'page-nav-button': PageNavButton,
  'site-nav-button': SiteNavButton,
};

const NAVBAR_CONTENT = `
<a slot="brand" href="#" title="Home" class="navbar-brand">Your Logo</a>
  <li><a href="#" class="nav-link">Topic 1</a></li>
  <li><a href="#" class="nav-link">Topic 2</a></li>
</a>
`;

const OMIT_PAGE_AND_SITE_NAV_BUTTONS = '<div></div>';

const SITE_NAV_BUTTON = `
<div class="nav-menu-container">
  <site-nav-button />
</div>
`;

const PAGE_NAV_BUTTON = `
<div class="nav-menu-container">
  <page-nav-button />
</div>
`;

const SITE_AND_PAGE_NAV_BUTTONS = `
<div class="nav-menu-container">
  <site-nav-button />
  <page-nav-button />
</div>
`;

describe('Navbar and secondary navbar', () => {
  test('navbar without site and page nav buttons', async () => {
    const wrapper = mount(Navbar, {
      slots: {
        default: NAVBAR_CONTENT,
        'lower-navbar': OMIT_PAGE_AND_SITE_NAV_BUTTONS,
      },
      stubs: DEFAULT_STUBS,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('navbar with site nav button', async () => {
    const wrapper = mount(Navbar, {
      slots: {
        default: NAVBAR_CONTENT,
        'lower-navbar': SITE_NAV_BUTTON,
      },
      stubs: DEFAULT_STUBS,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('navbar with page nav button', async () => {
    const wrapper = mount(Navbar, {
      slots: {
        default: NAVBAR_CONTENT,
        'lower-navbar': PAGE_NAV_BUTTON,
      },
      stubs: DEFAULT_STUBS,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('navbar with site and page nav buttons', async () => {
    const wrapper = mount(Navbar, {
      slots: {
        default: NAVBAR_CONTENT,
        'lower-navbar': SITE_AND_PAGE_NAV_BUTTONS,
      },
      stubs: DEFAULT_STUBS,
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
