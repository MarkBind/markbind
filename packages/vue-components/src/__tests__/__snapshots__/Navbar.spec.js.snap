// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar and secondary navbar navbar with page nav button 1`] = `
<div>
  <nav
    class="navbar navbar-expand-md navbar-dark bg-primary"
  >
    <div
      class="container-fluid"
    >
      <div
        class="navbar-left"
      >
        <a
          class="navbar-brand"
          href="#"
          title="Home"
        >
          Your Logo
        </a>
      </div>
       
      <div
        class="navbar-default"
      >
        <ul
          class="navbar-nav me-auto mt-2 mt-lg-0"
        >
          <li
            class="current"
          >
            <a
              class="nav-link"
              href="#"
            >
              Topic 1
            </a>
          </li>
           
          <li>
            <a
              class="nav-link"
              href="#"
            >
              Topic 2
            </a>
          </li>
        </ul>
      </div>
       
      <!---->
    </div>
  </nav>
   
  <div
    class="lower-navbar-container"
    style="display: none;"
  >
    <div
      class="nav-menu-container"
    >
      <!---->
    </div>
  </div>
</div>
`;

exports[`Navbar and secondary navbar navbar with site and page nav buttons 1`] = `
<div>
  <nav
    class="navbar navbar-expand-md navbar-dark bg-primary"
  >
    <div
      class="container-fluid"
    >
      <div
        class="navbar-left"
      >
        <a
          class="navbar-brand"
          href="#"
          title="Home"
        >
          Your Logo
        </a>
      </div>
       
      <div
        class="navbar-default"
      >
        <ul
          class="navbar-nav me-auto mt-2 mt-lg-0"
        >
          <li
            class="current"
          >
            <a
              class="nav-link"
              href="#"
            >
              Topic 1
            </a>
          </li>
           
          <li>
            <a
              class="nav-link"
              href="#"
            >
              Topic 2
            </a>
          </li>
        </ul>
      </div>
       
      <!---->
    </div>
  </nav>
   
  <div
    class="lower-navbar-container"
    style="display: none;"
  >
    <div
      class="nav-menu-container"
    >
      <!---->
       
      <!---->
    </div>
  </div>
</div>
`;

exports[`Navbar and secondary navbar navbar with site nav button 1`] = `
<div>
  <nav
    class="navbar navbar-expand-md navbar-dark bg-primary"
  >
    <div
      class="container-fluid"
    >
      <div
        class="navbar-left"
      >
        <a
          class="navbar-brand"
          href="#"
          title="Home"
        >
          Your Logo
        </a>
      </div>
       
      <div
        class="navbar-default"
      >
        <ul
          class="navbar-nav me-auto mt-2 mt-lg-0"
        >
          <li
            class="current"
          >
            <a
              class="nav-link"
              href="#"
            >
              Topic 1
            </a>
          </li>
           
          <li>
            <a
              class="nav-link"
              href="#"
            >
              Topic 2
            </a>
          </li>
        </ul>
      </div>
       
      <!---->
    </div>
  </nav>
   
  <div
    class="lower-navbar-container"
    style="display: none;"
  >
    <div
      class="nav-menu-container"
    >
      <!---->
    </div>
  </div>
</div>
`;

exports[`Navbar and secondary navbar navbar without site and page nav buttons 1`] = `
<div>
  <nav
    class="navbar navbar-expand-md navbar-dark bg-primary"
  >
    <div
      class="container-fluid"
    >
      <div
        class="navbar-left"
      >
        <a
          class="navbar-brand"
          href="#"
          title="Home"
        >
          Your Logo
        </a>
      </div>
       
      <div
        class="navbar-default"
      >
        <ul
          class="navbar-nav me-auto mt-2 mt-lg-0"
        >
          <li
            class="current"
          >
            <a
              class="nav-link"
              href="#"
            >
              Topic 1
            </a>
          </li>
           
          <li>
            <a
              class="nav-link"
              href="#"
            >
              Topic 2
            </a>
          </li>
        </ul>
      </div>
       
      <!---->
    </div>
  </nav>
   
  <div
    class="lower-navbar-container"
    style="display: none;"
  >
    <div />
  </div>
</div>
`;
